From 10325422e5a4735aa87a4d47bdb377e09fe963d2 Mon Sep 17 00:00:00 2001
From: Andy Klimczak <andyklimczak@fastmail.com>
Date: Sat, 22 Oct 2016 21:54:14 -0400
Subject: [PATCH] LANG-1276: StrBuilder#replaceAll
 ArrayIndexOutOfBoundsException (closes #200)

Avoid ArrayIndexOutOfBoundsException by keeping variable buf consistent with buffer in StrBuilder#replaceImpl.
---
 src/main/java/org/apache/commons/lang3/text/StrBuilder.java   | 2 +-
 .../java/org/apache/commons/lang3/text/StrBuilderTest.java    | 4 ++++
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/main/java/org/apache/commons/lang3/text/StrBuilder.java b/src/main/java/org/apache/commons/lang3/text/StrBuilder.java
index 5679484ab59..6079101cda0 100644
--- a/src/main/java/org/apache/commons/lang3/text/StrBuilder.java
+++ b/src/main/java/org/apache/commons/lang3/text/StrBuilder.java
@@ -2110,8 +2110,8 @@ private StrBuilder replaceImpl(
             return this;
         }
         final int replaceLen = (replaceStr == null ? 0 : replaceStr.length());
-        final char[] buf = buffer;
         for (int i = from; i < to && replaceCount != 0; i++) {
+            final char[] buf = buffer;
             final int removeLen = matcher.isMatch(buf, i, from, to);
             if (removeLen > 0) {
                 replaceImpl(i, i + removeLen, removeLen, replaceStr, replaceLen);
diff --git a/src/test/java/org/apache/commons/lang3/text/StrBuilderTest.java b/src/test/java/org/apache/commons/lang3/text/StrBuilderTest.java
index 7fa05794220..f83dc491201 100644
--- a/src/test/java/org/apache/commons/lang3/text/StrBuilderTest.java
+++ b/src/test/java/org/apache/commons/lang3/text/StrBuilderTest.java
@@ -897,6 +897,10 @@ public void testReplaceAll_StrMatcher_String() {
         sb = new StrBuilder("A1-A2A3-A4");
         sb.replaceAll(A_NUMBER_MATCHER, "***");
         assertEquals("***-******-***", sb.toString());
+
+        sb = new StrBuilder("Dear X, hello X.");
+        sb.replaceAll(StrMatcher.stringMatcher("X"), "012345678901234567");
+        assertEquals("Dear 012345678901234567, hello 012345678901234567.", sb.toString());
     }
 
     @Test
